### ISSUE TITLE
[Ch 01-04][Notebook][Major] sklearn 챕터가 대응 BERT 챕터를 한 칸 앞으로 지칭 — 4개 챕터 7곳

### ISSUE THREAD

**Issue by modesta-jihye-lee** [yoon-gu/neuqes-101#83]
## 대상

- **챕터**: Ch 01 · Ch 02 · Ch 03 · Ch 04 (sklearn 구간 전체)
- **산출물**: Notebook 4개 — `01_tfidf` `cell021` / `02_sklearn_regression` `cell004`·`cell018` / `03_sklearn_binary` `cell004`·`cell019` / `04_softmax_binary` `cell016`
- **기준 커밋**: `eed3dbe`
- **환경**: 정적 검토 + README 변화추적표(단일 출처) 및 Ch 7~12 노트북 대조. 로컬 재실행(sklearn 1.9.0) 4개 챕터 모두 에러 0건. **Ch 01~04 Colab 전체 실행 완료** (2026-08-15 · CPU 런타임 · 63/80/43/45초 · 모두 에러 0건)

> **한 이슈로 묶은 이유**: 네 챕터의 결함이 같은 유형(대응 BERT 챕터를 한 칸 앞으로 지칭)·같은 뿌리(`3f6f200` 재생성 회귀)·같은 수정 경로(마크다운 셀 → 재변환 → 재빌드)입니다. 담당자가 아래 표 하나로 7곳을 한 번에 고치고, 재검수자도 한 번에 확인할 수 있습니다. 챕터별 추적은 표의 챕터 열과 연결된 챕터 이슈 #25·#26·#27·#28 로 유지됩니다.

## 실제 결과

sklearn 구간(Ch 1-4)의 네 챕터가 **대응되는 BERT 챕터를 일관되게 한 칸 앞으로** 가리킵니다. 총 **7곳**입니다.

| # | 챕터 | 위치 | 현재 표기 | 실제 |
|---|---|---|---|---|
| 1 | Ch 01 | `cell021` 라이브러리 표 `datasets` | "**Ch 7**에서 깊게 본다" | **Ch 8** |
| 2 | Ch 02 | `cell004` Loss 노트 코드 주석 | `# PyTorch (Ch 8 이후 등장)` | **Ch 9** |
| 3 | Ch 02 | `cell018` 라이브러리 표 `LinearRegression` | "BERT는 **Ch 8**부터 같은 역할" | **Ch 9** |
| 4 | Ch 02 | `cell018` 라이브러리 표 `mean_squared_error` | "**Ch 8** BERT 회귀에서도 평가 지표로 등장" | **Ch 9** |
| 5 | Ch 03 | `cell004` `BCEWithLogitsLoss` 스니펫 주석 | `# PyTorch (Ch 9 이후)` | **Ch 10** |
| 6 | Ch 03 | `cell019` 라이브러리 표 `log_loss` | "**Ch 9** 이후 BERT binary에서도" | **Ch 10·11** |
| 7 | Ch 04 | `cell016` 라이브러리 표 `LogisticRegression` | "**Ch 11** BERT multi-class 에서 같은 패러다임" | **Ch 12** |

### 근거 — README 변화추적표(단일 출처)

| 챕터 | README 표 | 실제 내용 |
|---|---|---|
| Ch 7 | 데이터 = "간단 영어 예시 문장" | `pipeline` 첫 만남. `datasets` 미사용 |
| **Ch 8** | 데이터 = "Yelp (**datasets 라이브러리 해부**)" · Loss 열 `—` | Tokenizer 옵션 + `datasets`. **학습·Loss 없음** |
| **Ch 9** | `DistilBERT 파인튜닝` · Loss **`MSELoss`** | 별점 1-5 회귀 |
| **Ch 10** | Loss **`BCEWithLogitsLoss`** | DistilBERT binary — sigmoid (방식 A) |
| Ch 11 | Loss `CrossEntropyLoss` · Head `Linear(H,2)` | DistilBERT **binary** — softmax (방식 B) |
| **Ch 12** | Loss `CrossEntropyLoss` · Head **`Linear(H,5)`** | DistilBERT **multi-class** — Yelp 5클래스 |

코드로도 교차 확인했습니다.

- `nn.MSELoss()` 를 쓰는 노트북: Ch 2 · **Ch 9** (Ch 8 에 없음)
- `BCEWithLogitsLoss()` 를 쓰는 노트북: Ch 3 · Ch 6 · **Ch 10** (Ch 9 에 없음)
- Ch 8 노트북 헤딩: `## 1. 🚀 datasets 로 Yelp 로드`, `## 3. 🛠️ datasets.map — 5,000건 일괄 토큰화`

### 같은 노트북 안에서 자기모순인 곳이 둘 있습니다

**Ch 03** 마지막 셀(`cell023`, 다음 챕터 예고)은 올바릅니다.

> 이 동등성 직관은 **Ch 10·11** 에서 BERT binary 의 두 방식을 비교할 때 그대로 재활용됩니다.

**Ch 04** 의 다른 셀들도 Ch 11 을 정확히 "binary 방식 B" 로 씁니다(`cell004` 의 `# PyTorch (Ch 11, 방식 B)`, FAQ Q2·Q4·Q6). `cell016` 한 행만 어긋납니다.

### `PyTorch (Ch 8 이후 등장)` 은 이중으로 어긋납니다 (#2)

`import torch` 는 **Ch 7**(`07_bert_pipeline`)에서 이미 등장합니다. 이 주석이 가리키려던 건 "`nn.MSELoss` 를 직접 쓰는 시점" 으로 보이는데, 그것도 Ch 9 입니다.

## 기대 결과

| # | 수정안 |
|---|---|
| 1 | `Ch 8에서 깊게 본다` |
| 2 | `# PyTorch (Ch 9 BERT 회귀에서 등장)` |
| 3 | "sklearn 모델 라인업의 시작, BERT는 **Ch 9**부터 같은 역할" |
| 4 | "**Ch 9** BERT 회귀에서도 평가 지표로 등장" |
| 5 | `# PyTorch (Ch 10 BERT binary에서 등장)` |
| 6 | "**Ch 10·11** BERT binary 에서도" |
| 7 | "Ch 5 에서 K=5 로 확장, **Ch 12** BERT multi-class 에서 같은 패러다임" |

## 추가 확인 — 네 곳 모두 한 번 고쳐졌다가 되돌아온 값입니다

커밋 `c94f758`(2026-06-24, "WikiDocs 2차 검수: HIGH·MED 사실정확성 수정") 이 `pages/` 를 정정했으나, 다음 날 `3f6f200`(Ch 1–8·10 페이지 재생성)에서 **원천 노트북의 옛 값이 페이지를 덮어썼습니다.**

```
c94f758  - | `datasets` | ... | Ch 7에서 깊게 본다 |
         + | `datasets` | ... | Ch 8에서 깊게 본다 |
3f6f200  (재생성) → 다시 Ch 7
```

Ch 02(`-wrapup.md:5,7`)·Ch 04(`-wrapup.md`)도 같은 이력입니다. **페이지만 고치면 다음 재변환에서 그대로 되돌아오므로, 노트북·`book/*.tex` 를 함께 고쳐야 합니다.** 되돌림 현상 자체는 별도 이슈(전 챕터 · 페이지 재생성 회귀)로 등록했습니다.

## 함께 다듬을 것 — 결함 아님 (Ch 04 FAQ Q2)

같은 수정 사이클에 태울 수 있어 함께 적습니다. Ch 04 FAQ Q2 의 챕터 범위가 느슨합니다.

> 이 커리큘럼의 BERT 챕터(**Ch 9-14**)는 **방식 B가 기본**이라 …

| 챕터 | Loss | 방식 |
|---|---|---|
| 9 | `MSELoss` | 회귀 (방식 아님) |
| 10 | `BCEWithLogitsLoss` | **방식 A** |
| 11 | `CrossEntropyLoss` | 방식 B |
| 12 | `CrossEntropyLoss` | 방식 B 계열 |
| 13 | `BCEWithLogitsLoss` (per-label) | 방식 A 계열 |
| 14 | `BCEWithLogitsLoss + λ·MSELoss` | 방식 A 계열 + 보조 |

"딥러닝 표준이 방식 B" 라는 일반론은 맞지만 *이 커리큘럼의 Ch 9-14 가 방식 B 기본* 이라고 하기는 어렵습니다. "Ch 11·12 가 방식 B, Ch 10·13·14 는 방식 A 계열" 정도로 다듬으면 정확해집니다.

## 재현 절차

```bash
# 7곳 일괄 확인
grep -n "깊게 본다"        01_tfidf/01_tfidf.ipynb
grep -n "Ch 8"             02_sklearn_regression/02_sklearn_regression.ipynb
grep -n "Ch 9"             03_sklearn_binary/03_sklearn_binary.ipynb
grep -n "BERT multi-class" 04_softmax_binary/04_softmax_binary.ipynb

# 실제 챕터 확인 (README = 단일 출처)
grep -n "^| 7 \|^| 8 \|^| 9 \|^| 10 \|^| 11 \|^| 12 " README.md

# Loss 를 실제로 쓰는 노트북
grep -rl "MSELoss()"            --include="*.ipynb" . | grep -v executed
grep -rl "BCEWithLogitsLoss()"  --include="*.ipynb" . | grep -v executed

# 되돌림 이력
git show c94f758 -- pages/01-tfidf-wrapup.md pages/02-sklearn_regression-wrapup.md \
                    pages/04-softmax_binary-wrapup.md | head -40
```

## 로그

```
$ grep -rl "MSELoss()" --include="*.ipynb" . | grep -v executed
./02_sklearn_regression/02_sklearn_regression.ipynb
./09_bert_regression/09_bert_regression.ipynb                ← Ch 8 아님

$ grep -rl "BCEWithLogitsLoss()" --include="*.ipynb" . | grep -v executed
./03_sklearn_binary/03_sklearn_binary.ipynb
./06_sklearn_multilabel/06_sklearn_multilabel.ipynb
./10_bert_binary_sigmoid/10_bert_binary_sigmoid.ipynb        ← Ch 9 아님

$ grep -c "import torch" 07_bert_pipeline/07_bert_pipeline.ipynb
2                                                            ← torch 는 Ch 7 에 이미 등장

cell004(Ch03): # PyTorch (Ch 9 이후)
cell019(Ch03): | `sklearn.metrics.log_loss` | BCE 평가 | Ch 9 이후 BERT binary에서도 |
cell023(Ch03): - 이 동등성 직관은 Ch 10·11에서 …            ← 여기는 정확
```

## 전파 범위

세 산출물 모두에 있습니다 — 각각 직접 확인했습니다.

| 챕터 | Notebook | WikiDocs | book/*.tex |
|---|---|---|---|
| Ch 01 | `cell021` ❌ | `pages/01-tfidf-wrapup.md:5` ❌ (회귀) | `ch01_tfidf.tex:524` ❌ `\ref{ch:07}` |
| Ch 02 | `cell004`·`cell018` ❌ 3곳 | `02-sklearn_regression.md:57`, `-wrapup.md:5,7` ❌ (회귀) | `ch02_sklearn_regression.tex:538,540` ❌ |
| Ch 03 | `cell004`·`cell019` ❌ 2곳 | `03-sklearn_binary.md`, `-wrapup.md` ❌ | `ch03_sklearn_binary.tex` ❌ |
| Ch 04 | `cell016` ❌ | `04-softmax_binary-wrapup.md` ❌ (회귀) | `ch04_softmax_binary.tex` ❌ |

## 심각도

**Major** — 교차참조 어긋남. 가이드 3쪽의 "Ch 6·7·8 챕터 번호 off-by-one" 사례와 동일 유형입니다. 실행은 막지 않지만, 학습자가 "BERT 회귀는 Ch 8" 로 알고 넘어가면 Ch 8(토크나이저·`datasets`)에서 회귀를 찾다 헤맵니다.

## 재실행 여부

마크다운 셀 수정이므로 **노트북 재실행은 불필요**합니다. WikiDocs 재변환 → EPUB 재빌드가 필요합니다.

## 후속 — Ch 05·06 도 같은 패턴 확인 필요

네 챕터가 모두 같은 방향으로 어긋나 있어 sklearn 구간 전체의 문제로 보입니다. **Ch 05·06 검수 시 라이브러리 표와 Loss 노트 주석의 챕터 번호를 우선 확인**하고, 같은 유형이 나오면 이 이슈에 추가하겠습니다.

## 담당

- 수정 담당자: 미정
- 재검수자: 등록자 (@modesta-jihye-lee)
- 연결: 챕터 검수 이슈 #25 · #26 · #27 · #28

**Comment by yoon-gu** [comment]
PASS — sklearn 챕터가 대응 BERT 챕터를 한 칸 앞으로 지칭하던 7곳이 **7/7 정정**됐고, 옛 표기는 산출물 전체에서 0건입니다.

A2 후속 확인이던 Ch 05·06 확장분은 재현되지 않아(Ch 05 BERT 전방참조 0건, Ch 06 5건 전부 정확) off-by-one 은 Ch 01-04 국한으로 확정됐습니다.

— PR #289 (master `6986584`, 2026-09-18). `origin/master` 기준으로 노트북·`executed/`·`pages/`·`tex` 전파까지 확인했습니다.
